aten/src/ATen/mps/MPSAllocator.h

```diff
@@ -0,0 +1,242 @@
+
+  // interface exposed to at::Allocator
+  id<MTLBuffer> Malloc(size_t size, bool sharedStorage);
+  void Free(void* ptr);
+  void EmptyCache();
+  bool isSharedBuffer(void* ptr);
+
+  inline id<MTLDevice> Device() const { return m_device; }
+  void enable_debug_info() { m_enable_debug_info = true; }
+  bool debug_info_enabled() const { return m_enable_debug_info; }
+  void set_shared_storage_mode(bool useSharedStorage);
+
+private:
+  const id<MTLDevice> m_device;
+  std::mutex m_mutex;
+  // allocated buffers by device pointer
+  std::unordered_map<void*, BufferBlock*> m_allocated_buffers;
+  // unallocated cached buffers larger than 1 MB
+  BufferPool m_large_pool_shared, m_large_pool_private;
+  // unallocated cached buffers 1 MB or smaller
+  BufferPool m_small_pool_shared, m_small_pool_private;
+  // total memory allocated by HeapAllocator
+  size_t m_total_allocated_memory;
+  // max buffer size allowed by Metal
+  size_t m_max_buffer_size;
+  // sets a soft upper bound to limit the total allocations
+  bool m_set_fraction;
+  // use "PYTORCH_DEBUG_MPS_ALLOCATOR" env-var to enable debug info
+  bool m_enable_debug_info;
+
+  HeapBlock* get_free_heap(AllocParams& p);
+  bool get_free_buffer(AllocParams& p);
+  BufferBlock* get_allocated_buffer_block(void* ptr);
+  bool alloc_buffer(AllocParams& p);
+  void free_buffer(BufferBlock* buffer_block);
+  void release_buffer(BufferBlock* buffer_block, bool remove_empty_heap = true);
+  void release_buffers(BufferPool& pool);
+  bool release_available_cached_buffers(const AllocParams& p);
+  bool release_cached_buffers();
+
+  BufferPool& get_pool(size_t Size, bool useShared) {
+      return Size <= kMaxSmallAlloc ? (useShared ? m_small_pool_shared : m_small_pool_private) :
+                                      (useShared ? m_large_pool_shared : m_large_pool_private);
+  }
+
+  size_t get_allocation_size(size_t Length, bool useShared) {
+    MTLSizeAndAlign sizeAlign = [m_device heapBufferSizeAndAlignWithLength:Length
+                                                                   options:HeapBlock::getOptions(useShared)];
+    return BufferBlock::alignUp(sizeAlign.size, sizeAlign.align);
+  }
+  // TODO: make this configurable
+  static size_t max_split_size() { return std::numeric_limits<size_t>::max(); }
+  // maximum size of device memory available for allocation in current process
+  size_t max_available_size() const { return [m_device recommendedMaxWorkingSetSize] - [m_device currentAllocatedSize]; }
+
+  static std::string format_size(uint64_t size) {
+    std::ostringstream os;
+    os.precision(2);
+    os << std::fixed;
+    if (size <= 1024UL) { os << size << " bytes"; }
+    else if (size <= 1048576UL) { os << ((float) size / 1024.0) << " KB"; }
+    else if (size <= 1073741824UL) { os << ((float) size / 1048576.0) << " MB"; }
+    else { os << ((float) size / 1073741824.0) << " GB"; }
+    return os.str();
+  }
```

Comment:
really shouldn't be copy pasting this everywhere grumpy face

Reply:
Currently we are using the CUDACaching allocator design and values are same.  Although I agree, we can refactor a bunch of this. But I would request we do this post-merge as this will require touching CUDA code as well to adopt the new changes.